In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("sample.csv")
df

STATION                 DATE SOURCE  LATITUDE  LONGITUDE  ELEVATION  \
0     72228013876  2003-01-01T00:00:00      9  33.56556    -86.745      187.5   
1     72228013876  2003-01-01T00:53:00      C  33.56556    -86.745      187.5   
2     72228013876  2003-01-01T01:00:00      9  33.56556    -86.745      187.5   
3     72228013876  2003-01-01T01:53:00      C  33.56556    -86.745      187.5   
4     72228013876  2003-01-01T02:00:00      9  33.56556    -86.745      187.5   
...           ...                  ...    ...       ...        ...        ...   
1578  72228013876  2003-01-31T21:53:00      C  33.56556    -86.745      187.5   
1579  72228013876  2003-01-31T22:00:00      9  33.56556    -86.745      187.5   
1580  72228013876  2003-01-31T22:53:00      C  33.56556    -86.745      187.5   
1581  72228013876  2003-01-31T23:00:00      9  33.56556    -86.745      187.5   
1582  72228013876  2003-01-31T23:53:00      C  33.56556    -86.745      187.5   

                           NAME REPORT_TYPE CALL_SIGN QUALITY_CONTROL  ...  \
0     BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020  ...   
1     BIRMINGHAM AIRPORT, AL US       FM-15     BHM              V020  ...   
2     BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020  ...   
3     BIRMINGHAM AIRPORT, AL US       FM-15     BHM              V020  ...   
4     BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020  ...   
...                         ...         ...       ...             ...  ...   
1578  BIRMINGHAM AIRPORT, AL US       FM-15     BHM              V020  ...   
1579  BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020  ...   
1580  BIRMINGHAM AIRPORT, AL US       FM-15     BHM              V020  ...   
1581  BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020  ...   
1582  BIRMINGHAM AIRPORT, AL US       FM-15     BHM              V020  ...   

       MW2   MW3  MW4 MW5  OC1  OE1  OE2  OE3  \
0      NaN   NaN  NaN NaN  NaN  NaN  NaN  NaN   
1      NaN   NaN  NaN NaN  NaN  NaN  NaN  NaN   
2      NaN   NaN  NaN NaN  NaN  NaN  NaN  NaN   
3     45,5  63,5  NaN NaN  NaN  NaN  NaN  NaN   
4      NaN   NaN  NaN NaN  NaN  NaN  NaN  NaN   
...    ...   ...  ...  ..  ...  ...  ...  ...   
1578   NaN   NaN  NaN NaN  NaN  NaN  NaN  NaN   
1579   NaN   NaN  NaN NaN  NaN  NaN  NaN  NaN   
1580   NaN   NaN  NaN NaN  NaN  NaN  NaN  NaN   
1581   NaN   NaN  NaN NaN  NaN  NaN  NaN  NaN   
1582   NaN   NaN  NaN NaN  NaN  NaN  NaN  NaN   

                                                    REM  \
0                                                   NaN   
1     MET084KBHM 010053Z 14010KT 10SM BKN050 BKN070 ...   
2                                                   NaN   
3     MET105KBHM 010153Z 14011KT 2 1/2SM RA BR BKN02...   
4                                                   NaN   
...                                                 ...   
1578  MET068KBHM 312153Z 26010KT 10SM CLR 13/M02 A30...   
1579                                                NaN   
1580  MET071KBHM 312253Z 28014KT 10SM OVC042 11/M01 ...   
1581                                                NaN   
1582  MET101METAR KBHM 312353Z 27010KT 10SM BKN034 O...   

                                                   EQD  
0                                                  NaN  
1                     Q01+000000PRSWM2N01 00000JPWTH 0  
2                                                  NaN  
3     Q01+000003PRSWM1Q02    003PRSWM1N01 02170JPWTH 0  
4                                                  NaN  
...                                                ...  
1578  Q01+000000PRSWM2Q02    003SCOTCVN01 00000JPWTH 0  
1579                                               NaN  
1580                  Q01+000000PRSWM2N01 00000JPWTH 0  
1581                                               NaN  
1582                  Q01+000000PRSWM2N01 00000JPWTH 0  

[1583 rows x 78 columns]

In [4]:
df.columns

Index(['STATION', 'DATE', 'SOURCE', 'LATITUDE', 'LONGITUDE', 'ELEVATION',
       'NAME', 'REPORT_TYPE', 'CALL_SIGN', 'QUALITY_CONTROL', 'WND', 'CIG',
       'VIS', 'TMP', 'DEW', 'SLP', 'AA1', 'AA2', 'AA3', 'AB1', 'AD1', 'AE1',
       'AH1', 'AH2', 'AH3', 'AH4', 'AH5', 'AH6', 'AI1', 'AI2', 'AI3', 'AI4',
       'AI5', 'AI6', 'AJ1', 'AL1', 'AN1', 'AT1', 'AT2', 'AT3', 'AT4', 'AT5',
       'AT6', 'GA1', 'GD1', 'GD2', 'GD3', 'GD4', 'GF1', 'GJ1', 'GK1', 'GP1',
       'GQ1', 'GR1', 'KA1', 'KA2', 'KB1', 'KB2', 'KB3', 'KD1', 'KD2', 'KE1',
       'MA1', 'MD1', 'MG1', 'MH1', 'MK1', 'MW1', 'MW2', 'MW3', 'MW4', 'MW5',
       'OC1', 'OE1', 'OE2', 'OE3', 'REM', 'EQD'],
      dtype='object')

In [8]:
df[['TMP', 'DEW', 'WND', 'AA1']].head(10)

TMP      DEW             WND          AA1
0  +9999,9  +9999,9  999,9,9,9999,9          NaN
1  +0150,5  +0122,5  140,5,N,0052,5  01,0000,9,5
2  +9999,9  +9999,9  999,9,9,9999,9          NaN
3  +0144,5  +0122,5  140,5,N,0057,5  01,0000,9,5
4  +9999,9  +9999,9  999,9,9,9999,9          NaN
5  +0139,5  +0128,5  170,5,N,0062,5  01,0000,9,5
6  +9999,9  +9999,9  999,9,9,9999,9          NaN
7  +0140,1  +0120,1  160,1,N,0041,1  01,0000,2,1
8  +0139,5  +0117,5  150,5,N,0041,5  01,0000,9,5
9  +9999,9  +9999,9  999,9,9,9999,9          NaN

In [9]:
df.describe()

STATION      LATITUDE     LONGITUDE  ELEVATION  AT6  GA1  GD4  MW5
count  1.583000e+03  1.583000e+03  1.583000e+03     1583.0  0.0  0.0  0.0  0.0
mean   7.222801e+10  3.356556e+01 -8.674500e+01      187.5  NaN  NaN  NaN  NaN
std    0.000000e+00  7.107673e-15  1.421535e-14        0.0  NaN  NaN  NaN  NaN
min    7.222801e+10  3.356556e+01 -8.674500e+01      187.5  NaN  NaN  NaN  NaN
25%    7.222801e+10  3.356556e+01 -8.674500e+01      187.5  NaN  NaN  NaN  NaN
50%    7.222801e+10  3.356556e+01 -8.674500e+01      187.5  NaN  NaN  NaN  NaN
75%    7.222801e+10  3.356556e+01 -8.674500e+01      187.5  NaN  NaN  NaN  NaN
max    7.222801e+10  3.356556e+01 -8.674500e+01      187.5  NaN  NaN  NaN  NaN

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1583 entries, 0 to 1582
Data columns (total 78 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   STATION          1583 non-null   int64  
 1   DATE             1583 non-null   object 
 2   SOURCE           1583 non-null   object 
 3   LATITUDE         1583 non-null   float64
 4   LONGITUDE        1583 non-null   float64
 5   ELEVATION        1583 non-null   float64
 6   NAME             1583 non-null   object 
 7   REPORT_TYPE      1583 non-null   object 
 8   CALL_SIGN        1583 non-null   object 
 9   QUALITY_CONTROL  1583 non-null   object 
 10  WND              1583 non-null   object 
 11  CIG              1583 non-null   object 
 12  VIS              1583 non-null   object 
 13  TMP              1583 non-null   object 
 14  DEW              1583 non-null   object 
 15  SLP              1583 non-null   object 
 16  AA1              856 non-null    object 
 17  AA2           

In [13]:
columns = [
    "STATION",
    "DATE",
    "SOURCE",
    "LATITUDE",
    "LONGITUDE",
    "ELEVATION",
    "NAME",
    "REPORT_TYPE",
    "CALL_SIGN",
    "QUALITY_CONTROL",
    "WND",
    "CIG",
    "VIS",
    "TMP",
    "DEW",
    "SLP"
]

In [15]:
df[columns].head(10)

STATION                 DATE SOURCE  LATITUDE  LONGITUDE  ELEVATION  \
0  72228013876  2003-01-01T00:00:00      9  33.56556    -86.745      187.5   
1  72228013876  2003-01-01T00:53:00      C  33.56556    -86.745      187.5   
2  72228013876  2003-01-01T01:00:00      9  33.56556    -86.745      187.5   
3  72228013876  2003-01-01T01:53:00      C  33.56556    -86.745      187.5   
4  72228013876  2003-01-01T02:00:00      9  33.56556    -86.745      187.5   
5  72228013876  2003-01-01T02:53:00      C  33.56556    -86.745      187.5   
6  72228013876  2003-01-01T03:00:00      9  33.56556    -86.745      187.5   
7  72228013876  2003-01-01T03:16:00      4  33.56556    -86.745      187.5   
8  72228013876  2003-01-01T03:53:00      C  33.56556    -86.745      187.5   
9  72228013876  2003-01-01T04:00:00      9  33.56556    -86.745      187.5   

                        NAME REPORT_TYPE CALL_SIGN QUALITY_CONTROL  \
0  BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020   
1  BIRMINGHAM AIRPORT, AL US       FM-15     BHM              V020   
2  BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020   
3  BIRMINGHAM AIRPORT, AL US       FM-15     BHM              V020   
4  BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020   
5  BIRMINGHAM AIRPORT, AL US       FM-15     BHM              V020   
6  BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020   
7  BIRMINGHAM AIRPORT, AL US       FM-16     KBHM             V020   
8  BIRMINGHAM AIRPORT, AL US       FM-15     BHM              V020   
9  BIRMINGHAM AIRPORT, AL US       NSRDB     99999            V020   

              WND          CIG           VIS      TMP      DEW      SLP  
0  999,9,9,9999,9  99999,9,9,9  999999,9,9,9  +9999,9  +9999,9  99999,9  
1  140,5,N,0052,5  01524,5,M,N  016000,5,N,1  +0150,5  +0122,5  10038,5  
2  999,9,9,9999,9  99999,9,9,9  999999,9,9,9  +9999,9  +9999,9  99999,9  
3  140,5,N,0057,5  00762,5,M,N  004000,5,N,1  +0144,5  +0122,5  10037,5  
4  999,9,9,9999,9  99999,9,9,9  999999,9,9,9  +9999,9  +9999,9  99999,9  
5  170,5,N,0062,5  00427,5,M,N  004800,5,N,1  +0139,5  +0128,5  10037,5  
6  999,9,9,9999,9  99999,9,9,9  999999,9,9,9  +9999,9  +9999,9  99999,9  
7  160,1,N,0041,1  02400,1,9,N  016093,1,N,1  +0140,1  +0120,1  99999,9  
8  150,5,N,0041,5  03353,5,M,N  016000,5,N,1  +0139,5  +0117,5  10037,5  
9  999,9,9,9999,9  99999,9,9,9  999999,9,9,9  +9999,9  +9999,9  99999,9

Features that I would pick + encoding method:

TIME: 

* DATE: Obtain hour and day_of_year from it. Perform cyclic encoding (sin and cos) on both hour and day_of_year

LOCATION: 

* LATITUDE and LONGITUDE: Perform Spherical 3D Embedding on them (change them to cartesian coordinates on the map):
```
phi = np.deg2rad(latitude); lam = np.deg2rad(longitude)
x = np.cos(phi) * np.cos(lam)
y = np.cos(phi) * np.sin(lam)
z = np.sin(phi)
```

* ELEVATION: No transformation

OTHER:

Wind Condtion: 
* WND: Only take 4th value (which is the wind speed in tenths of m/s, so 0057 is 5.7 m/s, do this transformation appropriately), make sure to only select values where the 5th value is either 1 or 5.

Cloud Cover/Ceiling
* CIG: Take 1st value which represents the cloud ceiling height, only use value where second value is 1 or 5 (also first value 01524 is 1524 m)

Visibility 
* VIS: Take 1st value which represents the horizontal visibility distance (used to represent haze and fogs), only use value where second value is 1 or 5 (also first value 016000 is 16000 m)

Dew Point
* DEW: Take 1st value which represents the dew point which represents atmospheric moisture, only use value where second value is 1 or 5 (also the +0122 actually means 12.2, look at the signs too)

Sea Level Pressure:
* SLP: Take 1st value which represents sea-level pressure in hPa with one decimal place (so 10038 is 1003.8 hPa), only use it if second value is 1 or 5. 
